File: backend/src/features/users/users.service.ts
import { pool } from "../../db";

export const getUserById = async (userId: string) => {
  const result = await pool.query(
    `SELECT id, email, created_at FROM users WHERE id = $1`,
    [userId]
  );
  if (result.rowCount === 0) throw new Error("User not found");
  return result.rows[0];
};

export const getPublicUser = async (userId: string) => {
  const result = await pool.query(
    `SELECT user_id AS id, first_name, last_name, profile_picture
     FROM profiles WHERE user_id = $1`,
    [userId]
  );
  if (result.rowCount === 0) throw new Error("User not found");
  return result.rows[0];
};

export const getUserProfileData = async (userId: string) => {
  const result = await pool.query(
    `SELECT user_id AS id, first_name, last_name, bio, gender, birthdate, city, looking_for
     FROM profiles WHERE user_id = $1`,
    [userId]
  );
  if (result.rowCount === 0) throw new Error("User not found");
  return result.rows[0];
};

export const getUserBioData = async (userId: string) => {
  const result = await pool.query(
    `SELECT user_id AS id, interests, hobbies, music_preferences, food_preferences,
            personality_traits, city, birthdate
     FROM profiles WHERE user_id = $1`,
    [userId]
  );
  if (result.rowCount === 0) throw new Error("User not found");
  return result.rows[0];
};

File: backend-java/src/main/java/com/nexus/features/users/UsersService.java
package com.nexus.features.users;

import org.springframework.dao.EmptyResultDataAccessException;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;

import java.util.Map;
import java.util.NoSuchElementException;
import java.util.UUID;

@Service
public class UsersService {

    private final JdbcTemplate db;

    public UsersService(JdbcTemplate db) {
        this.db = db;
    }

    /** /users/:id — name + profile picture */
    public Map<String, Object> getPublicUser(UUID userId) {
        try {
            return db.queryForMap(
                "SELECT user_id AS id, first_name, last_name, profile_picture FROM profiles WHERE user_id = ?",
                userId
            );
        } catch (EmptyResultDataAccessException e) {
            throw new NoSuchElementException("User not found");
        }
    }

    /** /users/:id/profile — about me info */
    public Map<String, Object> getUserProfileData(UUID userId) {
        try {
            return db.queryForMap(
                """
                SELECT user_id AS id, first_name, last_name, bio, gender,
                       birthdate, city
                FROM profiles WHERE user_id = ?
                """,
                userId
            );
        } catch (EmptyResultDataAccessException e) {
            throw new NoSuchElementException("User not found");
        }
    }

    /** /users/:id/bio — biographical data for recommendations */
    public Map<String, Object> getUserBioData(UUID userId) {
        try {
            return db.queryForMap(
                """
                SELECT user_id AS id, interests, hobbies, music_preferences,
                       food_preferences, personality_traits, city, birthdate
                FROM profiles WHERE user_id = ?
                """,
                userId
            );
        } catch (EmptyResultDataAccessException e) {
            throw new NoSuchElementException("User not found");
        }
    }
}